### PR TITLE
Add hooks to retrieve module on install/upgrade from external source

### LINKED ISSUE #27856: Retrieve module source from Marketplace module
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

_No response_

### Describe the solution you'd like

As any connection to addons is removed from the Core, this feature is willing to add hook to allow external marketplaces to retrieve source (ZIP file) to install or upgrade a module

### Alternatives you've considered

_No response_

### Additional context

_No response_

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add hooks to allow modules (marketplace) to retrieve source when installing or upgrading a specific module.
| Type?             | new feature
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #27856 and [#MBO-46](https://forge.prestashop.com/browse/MBO-46)
| How to test?      | Can be tested with https://github.com/PrestaShopCorp/ps_mbo/pull/132
| Possible impacts? | -

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27703)

## Breaking changes

- ModuleManager now have an extra parameter HookManager to execute hooks

### CODE REVIEW

File: src/Core/Addon/Module/ModuleManager.php
Comment (matks) [inline src/Core/Addon/Module/ModuleManager.php:301]: Question: rather than choosing the path "$source is empty" can we provide a $source$ that would be marketplace?

**Review by matks — CHANGES_REQUESTED** [review]
Friendly reminder @sowbiba 😄

File: src/Core/Addon/Module/ModuleManager.php
Comment (sowbiba) [reply src/Core/Addon/Module/ModuleManager.php:301]: I'm afraid I don't understand your question
You mean get the source from marketplace before calling `install` ? or give `$source = 'marketplace'` ?
I recall that the marketplace is a non-native module and may be missing

Comment (saulaski) [reply src/Core/Addon/Module/ModuleManager.php:301]: Isn't it the purpose of this empty condition ? i.e. triggering an installed module (like the marketplace) that will download the missing source ?

Now this makes me curious :)
What should happen if 2 or more modules handle to the same hook `actionAdminModuleInstallRetrieveSource` ?

**Review by saulaski — DISMISSED** [review]

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Could you open an issue on the project, please ?

File: src/Core/Addon/Module/ModuleManager.php
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleManager.php:132]: Could you document the BC Break ?

File: src/Core/Addon/Module/ModuleManager.php
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleManager.php]:
```suggestion
        // The module is not in modules folder and no ZIP is given as source
```

File: tests/Unit/Core/Addon/Module/ModuleManagerTest.php
Comment (Progi1984) [inline tests/Unit/Core/Addon/Module/ModuleManagerTest.php:104]: Please add PHPDoc

File: src/Core/Addon/Module/ModuleManager.php
Comment (sowbiba) [reply src/Core/Addon/Module/ModuleManager.php:132]: Done

**Comment by sowbiba** [comment]
To answer @saulaski : "What should happen if 2 or more modules handle to the same hook actionAdminModuleInstallRetrieveSource ?"

Now I handle when multiple sources are returned. The first one is taken

File: src/Core/Addon/Module/ModuleManager.php
Comment (saulaski) [inline src/Core/Addon/Module/ModuleManager.php:433]: It's ok, but I would do without an extra array and hard coded index 0.
Returns `false` if array is empty, as safety measure.
```suggestion
                $sourceZip = reset($sourceZip);
```
Don't ever trust an array in your home... they're evil :))

**Review by saulaski — APPROVED** [review]
Beside the small code optimization I propose, LGTM :)

File: src/Core/Addon/Module/ModuleManager.php
Comment (saulaski) [inline src/Core/Addon/Module/ModuleManager.php:437]: If strict mode was enabled in this PHP file, or simply executed with PHP 8, `is_file` would only accept a `string` as argument. Or trigger a PHP error otherwise.
If I was paranoid, I would add a quick validation first:
```suggestion
            if (is_string($sourceZip) && is_file($sourceZip)) {
```
Wait ! I am paranoid, always :))

**Review by saulaski — APPROVED** [review]

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@sowbiba Could you fix integration tests ?

**Comment by sowbiba** [comment]
This PR will be replaced by https://github.com/PrestaShop/PrestaShop/pull/27632 in most of its parts

I'm saving Core devs resources by closing it and will apply interesting parts of it once #27632 will be merged

Thanks all for the reviews. It's not lost at all ❤️

**Pull request closed without merging**
